### PR TITLE
Add the possibility to do 'openssl help [command]'

### PR DESCRIPTION
**Pull request by levitte**
(inspired from a tweet)

### CODE REVIEW

**Comment by dot-asm** [comment]
Red cross from Travis is relevant.

File: apps/openssl.c [hunk 1]
```diff
@@ -442,6 +442,8 @@ typedef enum HELP_CHOICE {
 } HELP_CHOICE;
 
 const OPTIONS help_options[] = {
+    {OPT_HELP_STR, OPT_hHELP, '-', "Usage: help [options]\n"},
+    {OPT_HELP_STR, OPT_hHELP, '-', "       help [command]\n"},
```
Comment (dot-asm) [inline apps/openssl.c]: Is duplication intentional?

Comment (levitte) [reply apps/openssl.c]: Yes. That's the way I figured out to make two lines

Comment (richsalz) [reply apps/openssl.c]: Look at other uses of OPT_HELP_STR, they all use 1 not OPT_hHELP

File: apps/openssl.c [hunk 2]
```diff
@@ -469,6 +471,10 @@ int help_main(int argc, char **argv)
         }
     }
 
+    if (opt_num_rest() == 1) {
+        char *new_argv[] = { opt_rest()[0], "--help", NULL };
```
Comment (richsalz) [inline apps/openssl.c]: blank line.  and I think this is more clear:
```
char *newargv[2];

newargv[0] = *opt_rest();
newargv[1] = "--help";
newargv[2] = NULL;

...
```
Do whatever, but I think dynamic-sized array initializers have been a problem on some platforms.

File: apps/openssl.c [hunk 3]
```diff
@@ -787,18 +793,23 @@ static void list_disabled(void)
...
     size_t i;
 
-    /* Sort alphabetically within category. For nicer help displays. */
-    for (i = 0, f = functions; f->name != NULL; ++f, ++i) ;
-    qsort(functions, i, sizeof(*functions), SortFnByName);
+    if (!prog_inited) {
+        prog_inited = 1;
+
+        /* Sort alphabetically within category. For nicer help displays. */
+        for (i = 0, f = functions; f->name != NULL; ++f, ++i) ;
```
Comment (richsalz) [inline apps/openssl.c]: please use continue.  or at least put the last semi-colon on its own line.

File: apps/openssl.c [hunk 4]
```diff
@@ -787,18 +793,23 @@ static void list_disabled(void)
...
-    LHASH_OF(FUNCTION) *ret;
+    static LHASH_OF(FUNCTION) *ret = NULL;
+    static int prog_inited = 0;
     FUNCTION *f;
     size_t i;
 
-    /* Sort alphabetically within category. For nicer help displays. */
-    for (i = 0, f = functions; f->name != NULL; ++f, ++i) ;
-    qsort(functions, i, sizeof(*functions), SortFnByName);
+    if (!prog_inited) {
```
Comment (richsalz) [inline apps/openssl.c]: you don't need prog_inited.  just do
``
if (ret != NULL)
    return ret;
```
And then don't indent the rest of the function body.

File: apps/openssl.c (hunk 1 above)
Comment (richsalz) [reply apps/openssl.c]: yes, but use 1 not OPT_hHELP the way all other instances of OPT_HELP_STR do.

File: apps/openssl.c (hunk 4 above)
Comment (levitte) [reply apps/openssl.c]: So if it fails creating the table, it should retry the next time this is called?  That flag is an indicator that it was tried once, don't even bother trying again

File: apps/openssl.c (hunk 2 above)
Comment (levitte) [reply apps/openssl.c]: It's actually fixed size, three elements.  But I do get complaints that initializing values dynamically isn't necessarely supported, so I still have to do it assignment style...

File: apps/openssl.c (hunk 4 above)
Comment (levitte) [reply apps/openssl.c]: ... however, I can still change the code around to avoid all that extra indentation ;-)

File: apps/openssl.c (hunk 3 above)
Comment (levitte) [reply apps/openssl.c]: Okie

File: apps/openssl.c (hunk 1 above)
Comment (levitte) [reply apps/openssl.c]: Okie, done

**Review by richsalz — APPROVED** [review]
looks good, thanks.

**Comment by levitte** [comment]
Ok!  Merged.  391e987e9cd837446d3d0bf5505bbc7aa44b627e and da6be198f7ab690b8ace83f35502bca8dd9218e5

**Pull request closed without merging**